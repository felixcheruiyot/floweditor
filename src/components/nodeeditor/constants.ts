export const DEFAULT_OPERAND = '@input';
export const GROUPS_OPERAND = '@contact.groups';
export const SUBFLOW_OPERAND = '@child.run.status';
export const WEBHOOK_OPERAND = '@run.webhook';

export const DEFAULT_BODY: string = `{
    "flow": @(json(run.flow)),
    "path": @(json(run.path)),
    "results": @(json(run.results)),
    "input": @(json(run.input)),
    "channel": @(json(run.input.channel))
    "contact": @(json(contact))
}`;
