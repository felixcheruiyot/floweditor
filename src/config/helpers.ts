import { FlowTypes, FlowTypeVisibility, Operator, Type } from '~/config/interfaces';

export const filterOperators = (operators: Operator[], flowType: FlowTypes): Operator[] => {
    return filterVisibility(operators, flowType);
};

export const filterTypeConfigs = (typeConfigs: Type[], flowType: FlowTypes): Type[] => {
    return filterVisibility(typeConfigs, flowType);
};

const filterVisibility = (items: FlowTypeVisibility[], flowType: FlowTypes): any[] => {
    return items.filter((item: FlowTypeVisibility) => {
        if (item.visibility === undefined) {
            return true;
        }

        return item.visibility.findIndex((ft: FlowTypes) => ft === flowType) > -1;
    });
};
