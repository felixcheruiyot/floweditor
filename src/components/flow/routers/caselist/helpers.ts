import { DraggableStyle } from 'react-beautiful-dnd';
import { CaseProps, DragCursor } from '~/components/flow/routers/caselist/CaseList';
import { Operators } from '~/config/interfaces';
import { getOperatorConfig } from '~/config/operatorConfigs';
import { createUUID } from '~/utils';

export const createEmptyCase = (): CaseProps => {
    const uuid = createUUID();
    const config = getOperatorConfig(Operators.has_any_word);
    return {
        uuid,
        kase: {
            uuid,
            type: config.type,
            arguments: [''],
            category_uuid: null
        },
        categoryName: '',
        valid: true
    };
};

export const getItemStyle = (
    draggableStyle: DraggableStyle,
    isDragging: boolean
): DraggableStyle => ({
    userSelect: 'none',
    outline: 'none',
    ...draggableStyle,
    // Overwriting default draggableStyle object from this point down
    ...(isDragging
        ? {
              background: '#f2f9fc',
              borderRadius: 4,
              opacity: 0.75,
              height: draggableStyle.height + 15
          }
        : {})
});

export const getListStyle = (isDraggingOver: boolean, single: boolean): { cursor: DragCursor } => {
    if (single) {
        return null;
    }

    return {
        cursor: isDraggingOver ? DragCursor.move : DragCursor.pointer
    };
};
