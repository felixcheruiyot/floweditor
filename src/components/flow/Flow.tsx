import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Canvas } from '~/components/canvas/Canvas';
import { CanvasDraggableProps } from '~/components/canvas/CanvasDraggable';
import Node from '~/components/flow/node/Node';
import { getDraggedFrom } from '~/components/helpers';
import NodeEditor from '~/components/nodeeditor/NodeEditor';
import Simulator from '~/components/simulator/Simulator';
import Sticky, { STICKY_BODY, STICKY_TITLE } from '~/components/sticky/Sticky';
import { ConfigProviderContext, fakePropType } from '~/config/ConfigProvider';
import { Exit, FlowDefinition } from '~/flowTypes';
import Plumber from '~/services/Plumber';
import editorState, { DragSelection, EditorState } from '~/store/editor';
import { RenderNode } from '~/store/flowContext';
import { detectLoops, getOrderedNodes } from '~/store/helpers';
import { NodeEditorSettings } from '~/store/nodeEditor';
import AppState from '~/store/state';
import {
    ConnectionEvent,
    DispatchWithState,
    ensureStartNode,
    EnsureStartNode,
    mergeEditorState,
    MergeEditorState,
    NoParamsAC,
    OnConnectionDrag,
    onConnectionDrag,
    OnOpenNodeEditor,
    onOpenNodeEditor,
    OnUpdateCanvasPositions,
    onUpdateCanvasPositions,
    resetNodeEditingState,
    UpdateConnection,
    updateConnection,
    updateSticky,
    UpdateSticky
} from '~/store/thunks';
import { contextTypes } from '~/testUtils';
import {
    ACTIVITY_INTERVAL,
    createUUID,
    isRealValue,
    NODE_PADDING,
    renderIf,
    snapToGrid,
    timeEnd,
    timeStart
} from '~/utils';
import Debug from '~/utils/debug';

export interface FlowStoreProps {
    editorState: Partial<EditorState>;
    mergeEditorState: MergeEditorState;

    definition: FlowDefinition;
    nodes: { [uuid: string]: RenderNode };
    dependencies: FlowDefinition[];
    nodeEditorSettings: NodeEditorSettings;

    ensureStartNode: EnsureStartNode;
    updateConnection: UpdateConnection;
    onOpenNodeEditor: OnOpenNodeEditor;
    onUpdateCanvasPositions: OnUpdateCanvasPositions;
    resetNodeEditingState: NoParamsAC;
    onConnectionDrag: OnConnectionDrag;
    updateSticky: UpdateSticky;
}

export interface Translations {
    [uuid: string]: any;
}

export const DRAG_THRESHOLD = 3;
export const REPAINT_TIMEOUT = 500;
export const GHOST_POSITION_INITIAL = { left: -1000, top: -1000 };

export const nodeSpecId = 'node';
export const nodesContainerSpecId = 'node-container';
export const ghostNodeSpecId = 'ghost-node';
export const dragSelectSpecId = 'drag-select';

export const isDraggingBack = (event: ConnectionEvent) => {
    return event.suspendedElementId === event.targetId && event.source !== null;
};

export const getDragStyle = (drag: DragSelection) => {
    const left = Math.min(drag.startX, drag.currentX);
    const top = Math.min(drag.startY, drag.currentY);
    const width = Math.max(drag.startX, drag.currentX) - left;
    const height = Math.max(drag.startY, drag.currentY) - top;
    return {
        left,
        top,
        width,
        height
    };
};

export class Flow extends React.Component<FlowStoreProps, {}> {
    private Plumber: Plumber;
    private ele: HTMLDivElement;
    private nodeContainerUUID: string;

    // Refs
    private ghost: any;

    public static contextTypes = {
        config: fakePropType
    };

    constructor(props: FlowStoreProps, context: ConfigProviderContext) {
        super(props, context);

        this.nodeContainerUUID = createUUID();

        this.Plumber = new Plumber();

        /* istanbul ignore next */
        if (context.config.debug) {
            window.fe = new Debug(props, this.props.editorState.debug);
        }

        bindCallbacks(this, {
            include: [/Ref$/, /^on/, /^is/, /^get/]
        });

        timeStart('RenderAndPlumb');
    }

    private onRef(ref: HTMLDivElement): HTMLDivElement {
        return (this.ele = ref);
    }

    private ghostRef(ref: any): any {
        return (this.ghost = ref);
    }

    public componentDidMount(): void {
        this.Plumber.bind('connection', (event: ConnectionEvent) =>
            this.props.updateConnection(event.sourceId, event.targetId)
        );
        this.Plumber.bind('beforeDrag', (event: ConnectionEvent) =>
            this.beforeConnectionDrag(event)
        );

        this.Plumber.bind('connectionDrag', (event: ConnectionEvent) => {
            this.props.onConnectionDrag(event, this.context.config.flowType);
        });

        this.Plumber.bind('connectionDragStop', (event: ConnectionEvent) =>
            this.onConnectorDrop(event)
        );
        this.Plumber.bind(
            'beforeStartDetach',
            (event: ConnectionEvent) => !this.props.editorState.translating
        );
        this.Plumber.bind('beforeDetach', (event: ConnectionEvent) => true);
        this.Plumber.bind('beforeDrop', (event: ConnectionEvent) =>
            this.onBeforeConnectorDrop(event)
        );

        // If we don't have any nodes, create our first one
        this.props.ensureStartNode();

        let offset = { left: 0, top: 0 };

        /* istanbul ignore next */
        if (this.ele) {
            offset = this.ele.getBoundingClientRect();
        }

        this.props.mergeEditorState({
            containerOffset: { left: offset.left, top: offset.top + window.scrollY }
        });

        timeEnd('RenderAndPlumb');

        // deals with safari load rendering throwing
        // off the jsplumb offsets
        window.setTimeout(() => this.Plumber.repaint(), REPAINT_TIMEOUT);
    }

    public componentWillUnmount(): void {
        this.Plumber.reset();
    }

    public componentWillUpdate(prevProps: FlowStoreProps): void {
        if (
            prevProps.editorState.activityInterval === this.props.editorState.activityInterval &&
            this.props.editorState.activityInterval !== ACTIVITY_INTERVAL
        ) {
            this.props.mergeEditorState({ activityInterval: ACTIVITY_INTERVAL });
        }
    }

    /**
     * Called right before a connector is dropped onto a new node
     */
    private onBeforeConnectorDrop(event: ConnectionEvent): boolean {
        this.props.resetNodeEditingState();
        const fromNodeUUID = event.sourceId.split(':')[0];
        try {
            detectLoops(this.props.nodes, fromNodeUUID, event.targetId);
        } catch {
            return false;
        }
        return true;
    }

    /**
     * Called the moment a connector is done dragging, whether it is dropped on an
     * existing node or on to empty space.
     */
    private onConnectorDrop(event: ConnectionEvent): boolean {
        const { ghostNode } = this.props.editorState;
        // Don't show the node editor if we a dragging back to where we were
        if (isRealValue(ghostNode) && !isDraggingBack(event)) {
            // Wire up the drag from to our ghost node
            this.Plumber.recalculate(ghostNode.node.uuid);

            const dragPoint = getDraggedFrom(ghostNode);
            this.Plumber.connect(
                dragPoint.nodeUUID + ':' + dragPoint.exitUUID,
                ghostNode.node.uuid
            );

            // Save our position for later
            const { left, top } = (this.ghost &&
                snapToGrid(
                    this.ghost.wrappedInstance.ele.offsetLeft,
                    this.ghost.wrappedInstance.ele.offsetTop
                )) || { left: 0, top: 0 };

            this.props.editorState.ghostNode.ui.position = { left, top };
            let originalAction = null;
            if (ghostNode.node.actions && ghostNode.node.actions.length === 1) {
                originalAction = ghostNode.node.actions[0];
            }

            // Bring up the node editor
            this.props.onOpenNodeEditor({
                originalNode: ghostNode,
                originalAction
            });
        }

        /* istanbul ignore next */
        document.removeEventListener('mousemove', (window as any).ghostListener);

        return true;
    }

    private beforeConnectionDrag(event: ConnectionEvent): boolean {
        return !this.props.editorState.translating;
    }

    private getNodes(): CanvasDraggableProps[] {
        return getOrderedNodes(this.props.nodes).map((renderNode: RenderNode, idx: number) => {
            return {
                uuid: renderNode.node.uuid,
                position: renderNode.ui.position,
                ele: (selected: boolean) => (
                    <Node
                        startingNode={idx === 0}
                        selected={selected}
                        key={renderNode.node.uuid}
                        data-spec={nodeSpecId}
                        nodeUUID={renderNode.node.uuid}
                        plumberMakeTarget={this.Plumber.makeTarget}
                        plumberRemove={this.Plumber.remove}
                        plumberRecalculate={this.Plumber.recalculate}
                        plumberMakeSource={this.Plumber.makeSource}
                        plumberConnectExit={this.Plumber.connectExit}
                        plumberUpdateClass={this.Plumber.updateClass}
                    />
                )
            };
        });
    }

    private getStickies(): CanvasDraggableProps[] {
        const stickyMap = this.props.definition._ui.stickies || {};
        return Object.keys(stickyMap).map(uuid => {
            return {
                uuid,
                ele: (selected: boolean) => (
                    <Sticky key={uuid} uuid={uuid} sticky={stickyMap[uuid]} selected={selected} />
                ),
                position: stickyMap[uuid].position
            };
        });
    }

    private getDragNode(): JSX.Element {
        return isRealValue(this.props.editorState.ghostNode) ? (
            <div style={{ position: 'absolute', display: 'block' }}>
                <Node
                    selected={false}
                    startingNode={false}
                    ref={this.ghostRef}
                    key={this.props.editorState.ghostNode.node.uuid}
                    data-spec={ghostNodeSpecId}
                    ghost={true}
                    nodeUUID={this.props.editorState.ghostNode.node.uuid}
                    plumberMakeTarget={this.Plumber.makeTarget}
                    plumberRemove={this.Plumber.remove}
                    plumberRecalculate={this.Plumber.recalculate}
                    plumberMakeSource={this.Plumber.makeSource}
                    plumberConnectExit={this.Plumber.connectExit}
                    plumberUpdateClass={this.Plumber.updateClass}
                />
            </div>
        ) : null;
    }

    private getSimulator(): JSX.Element {
        return renderIf(
            this.context.config.endpoints && this.context.config.endpoints.simulateStart
        )(<Simulator mergeEditorState={this.props.mergeEditorState} />);
    }

    private getNodeEditor(): JSX.Element {
        return renderIf(this.props.nodeEditorSettings !== null)(
            <NodeEditor
                plumberConnectExit={this.Plumber.connectExit}
                plumberRepaintForDuration={this.Plumber.repaintForDuration}
            />
        );
    }

    private isClickOnCanvas(event: React.MouseEvent<HTMLDivElement>): boolean {
        // TODO: not sure the TS-safe way to access id here
        return (event.target as any).id === this.nodeContainerUUID;
    }

    private onDoubleClick(event: React.MouseEvent<HTMLDivElement>): void {
        if (this.isClickOnCanvas(event)) {
            const { left, top } = snapToGrid(
                event.pageX - this.props.editorState.containerOffset.left - 100 + NODE_PADDING,
                event.pageY - this.props.editorState.containerOffset.top - NODE_PADDING * 2 - 40
            );

            this.props.updateSticky(createUUID(), {
                position: { left, top },
                title: STICKY_TITLE,
                body: STICKY_BODY
            });
        }
    }

    public render(): JSX.Element {
        const draggables = this.getStickies().concat(this.getNodes());

        return (
            <div onDoubleClick={this.onDoubleClick} ref={this.onRef}>
                <Canvas
                    onDragging={(uuids: string[]) => {
                        uuids.forEach((uuid: string) => {
                            if (uuid in this.props.nodes) {
                                this.props.nodes[uuid].node.exits.forEach((exit: Exit) => {
                                    if (exit.destination_uuid) {
                                        uuids.push(uuid + ':' + exit.uuid);
                                    }
                                });
                            }
                        });
                        this.Plumber.recalculateUUIDs(uuids);
                    }}
                    uuid={this.nodeContainerUUID}
                    dragActive={this.props.editorState.dragActive}
                    mergeEditorState={this.props.mergeEditorState}
                    draggables={draggables}
                    onUpdatePositions={this.props.onUpdateCanvasPositions}
                >
                    {this.getSimulator()}
                    {this.getDragNode()}
                    {this.getNodeEditor()}
                </Canvas>
            </div>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = ({
    flowContext: { definition, dependencies, nodes },
    editorState,
    nodeEditor: { settings }
}: AppState) => {
    return {
        nodeEditorSettings: settings,
        definition,
        nodes,
        dependencies,
        editorState: editorState as Partial<EditorState>
    };
};

/* istanbul ignore next */
const mapDispatchToProps = (dispatch: DispatchWithState) =>
    bindActionCreators(
        {
            mergeEditorState,
            ensureStartNode,
            resetNodeEditingState,
            onConnectionDrag,
            onOpenNodeEditor,
            onUpdateCanvasPositions,
            updateConnection,
            updateSticky
        },
        dispatch
    );

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Flow);
