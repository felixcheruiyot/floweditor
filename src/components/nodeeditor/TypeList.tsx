import * as React from 'react';
import Select from 'react-select';
import * as styles from '~/components/nodeeditor/TypeList.scss';
import { fakePropType } from '~/config/ConfigProvider';
import { filterTypeConfigs } from '~/config/helpers';
import { Type } from '~/config/interfaces';
import { configsToDisplay } from '~/config/typeConfigs';
import { large } from '~/utils/reactselect';

export interface TypeListProps {
    __className: string;
    initialType: Type;
    onChange(config: Type): void;
}

export interface TypeListState {
    config: Type;
}

export default class TypeList extends React.PureComponent<TypeListProps, TypeListState> {
    private typeConfigs: Type[];

    constructor(props: TypeListProps) {
        super(props);

        this.state = {
            config: this.props.initialType
        };

        this.handleChangeType = this.handleChangeType.bind(this);
    }

    public static contextTypes = {
        config: fakePropType
    };

    private handleChangeType(config: Type): void {
        this.setState(
            {
                config
            },
            () => this.props.onChange(config)
        );
    }

    private getTypeConfigs(): Type[] {
        if (this.typeConfigs === undefined) {
            this.typeConfigs = filterTypeConfigs(configsToDisplay, this.context.config.flowType);
        }
        return this.typeConfigs;
    }

    public render(): JSX.Element {
        return (
            <div className={`${this.props.__className} ${styles.typeList}`}>
                <p>When a contact arrives at this point in your flow...</p>
                <div>
                    <Select
                        styles={large}
                        value={this.state.config}
                        onChange={this.handleChangeType}
                        isSearchable={false}
                        isClearable={false}
                        getOptionValue={(option: Type) => option.type}
                        getOptionLabel={(option: Type) => option.description}
                        options={this.getTypeConfigs()}
                    />
                </div>
            </div>
        );
    }
}
