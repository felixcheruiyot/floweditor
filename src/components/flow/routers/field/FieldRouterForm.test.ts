import { RouterFormProps } from '~/components/flow/props';
import { CaseProps } from '~/components/flow/routers/caselist/CaseList';
import { DEFAULT_OPERAND } from '~/components/nodeeditor/constants';
import { Operators, Types } from '~/config/interfaces';
import { RouterTypes, SwitchRouter } from '~/flowTypes';
import { AssetType } from '~/store/flowContext';
import { composeComponentTestUtils, mock } from '~/testUtils';
import { createRenderNode, getRouterFormProps } from '~/testUtils/assetCreators';
import * as utils from '~/utils';

import FieldRouterForm from './FieldRouterForm';
import { createUUID } from '~/utils';

const { setup } = composeComponentTestUtils<RouterFormProps>(
    FieldRouterForm,
    getRouterFormProps(
        createRenderNode({
            actions: [],
            exits: [],
            ui: {
                position: { left: 0, top: 0 },
                type: Types.split_by_contact_field,
                config: {
                    router: {
                        operandAsset: {
                            id: 'favorite_color',
                            name: 'Favorite Color',
                            type: AssetType.Field
                        }
                    },
                    delimit: {
                        index: 0,
                        delimiter: ' '
                    }
                }
            }
        })
    )
);

mock(utils, 'createUUID', utils.seededUUIDs());

const exitOneUUID = utils.createUUID();
const redCaseUUID = utils.createUUID();

describe(FieldRouterForm.name, () => {
    it('should render', () => {
        const { wrapper } = setup(true);
        expect(wrapper).toMatchSnapshot();
    });

    it('initializes', () => {
        const { wrapper } = setup(true, {
            nodeSettings: {
                $set: {
                    originalNode: createRenderNode({
                        actions: [],
                        exits: [{ destination_uuid: null, uuid: exitOneUUID }],
                        router: {
                            type: RouterTypes.switch,
                            operand: DEFAULT_OPERAND,
                            categories: [
                                { uuid: createUUID(), name: 'Other', exit_uuid: exitOneUUID }
                            ],
                            cases: [
                                {
                                    uuid: redCaseUUID,
                                    type: Operators.has_any_word,
                                    arguments: ['red'],
                                    category_uuid: null
                                }
                            ],
                            default_category_uuid: exitOneUUID,
                            result_name: 'Color'
                        } as SwitchRouter,
                        ui: {
                            position: { left: 0, top: 0 },
                            type: Types.split_by_contact_field,
                            config: {
                                router: {
                                    operandAsset: {
                                        id: 'favorite_color',
                                        name: 'Favorite Color',
                                        type: AssetType.Field
                                    }
                                },
                                delimit: {
                                    delimiter: ' ',
                                    index: 0
                                }
                            }
                        }
                    })
                }
            }
        });

        expect(wrapper).toMatchSnapshot();
    });

    describe('updates', () => {
        it('should save changes', () => {
            const { instance, props } = setup(true, {
                $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
            });

            instance.handleUpdateResultName('Favorite Color');
            instance.handleCasesUpdated([
                { kase: { type: Operators.has_any_word, arguments: ['red'] }, categoryName: 'Red' },
                {
                    kase: { type: Operators.has_any_word, arguments: ['maroon'] },
                    categoryName: 'Red'
                },
                {
                    kase: { type: Operators.has_any_word, arguments: ['green'] },
                    categoryName: 'Green'
                }
            ] as CaseProps[]);

            expect(instance.state).toMatchSnapshot();

            instance.handleSave();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateRouter).toHaveBeenCalled();
            expect(props.updateRouter).toMatchCallSnapshot();
        });

        it('should cancel', () => {
            const { instance, props } = setup(true, {
                $merge: { updateRouter: jest.fn(), onClose: jest.fn() }
            });

            instance.handleFieldChanged({ id: 'viber', name: 'Viber', type: AssetType.URN });
            instance.getButtons().secondary.onClick();
            expect(props.onClose).toHaveBeenCalled();
            expect(props.updateRouter).not.toHaveBeenCalled();
        });
    });
});
