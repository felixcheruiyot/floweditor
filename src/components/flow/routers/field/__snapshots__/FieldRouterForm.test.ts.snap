// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldRouterForm initializes 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_contact_field"
  title="Split by Contact"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by contact",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Split by Contact",
        "type": "split_by_contact_field",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="leadIn"
  >
    If the contact's
    <div
      className="fieldSelect select-small"
    >
      <AssetSelector
        additionalOptions={
          Array [
            Object {
              "id": "name",
              "name": "Name",
              "type": "property",
            },
            Object {
              "id": "language",
              "name": "Language",
              "type": "property",
            },
            Object {
              "id": "channel",
              "name": "Channel",
              "type": "property",
            },
            Object {
              "id": "ext",
              "name": "External ID",
              "type": "scheme",
            },
            Object {
              "id": "facebook",
              "name": "Facebook ID",
              "type": "scheme",
            },
            Object {
              "id": "fcm",
              "name": "Firebase ID",
              "type": "scheme",
            },
            Object {
              "id": "jiochat",
              "name": "Jiochat ID",
              "type": "scheme",
            },
            Object {
              "id": "line",
              "name": "Line ID",
              "type": "scheme",
            },
            Object {
              "id": "mailto",
              "name": "Email Address",
              "type": "scheme",
            },
            Object {
              "id": "tel",
              "name": "Phone Number",
              "type": "scheme",
            },
            Object {
              "id": "telegram",
              "name": "Telegram ID",
              "type": "scheme",
            },
            Object {
              "id": "twitterid",
              "name": "Twitter ID",
              "type": "scheme",
            },
            Object {
              "id": "wechat",
              "name": "Wechat ID",
              "type": "scheme",
            },
            Object {
              "id": "whatsapp",
              "name": "Whatsapp Number",
              "type": "scheme",
            },
            Object {
              "id": "viber",
              "name": "Viber ID",
              "type": "scheme",
            },
          ]
        }
        assets={
          Object {
            "items": Object {},
            "type": "field",
          }
        }
        entry={
          Object {
            "value": Object {
              "id": "favorite_color",
              "name": null,
              "type": "field",
            },
          }
        }
        name="Contact Field"
        onChange={[Function]}
        searchable={true}
        sortFunction={[Function]}
        styles={
          Object {
            "control": [Function],
            "indicatorsContainer": [Function],
            "input": [Function],
            "menuList": [Function],
            "option": [Function],
            "singleValue": [Function],
            "valueContainer": [Function],
          }
        }
      />
    </div>
  </div>
  <CaseList
    cases={
      Array [
        Object {
          "categoryName": null,
          "kase": Object {
            "arguments": Array [],
            "category_uuid": null,
            "type": "has_any_word",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          "valid": true,
        },
      ]
    }
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "Color",
      }
    }
  />
</Dialog>
`;

exports[`FieldRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_contact_field"
  title="Split by Contact"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by contact",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
          "cases",
        ],
        "name": "Split by Contact",
        "type": "split_by_contact_field",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="leadIn"
  >
    If the contact's
    <div
      className="fieldSelect select-small"
    >
      <AssetSelector
        additionalOptions={
          Array [
            Object {
              "id": "name",
              "name": "Name",
              "type": "property",
            },
            Object {
              "id": "language",
              "name": "Language",
              "type": "property",
            },
            Object {
              "id": "channel",
              "name": "Channel",
              "type": "property",
            },
            Object {
              "id": "ext",
              "name": "External ID",
              "type": "scheme",
            },
            Object {
              "id": "facebook",
              "name": "Facebook ID",
              "type": "scheme",
            },
            Object {
              "id": "fcm",
              "name": "Firebase ID",
              "type": "scheme",
            },
            Object {
              "id": "jiochat",
              "name": "Jiochat ID",
              "type": "scheme",
            },
            Object {
              "id": "line",
              "name": "Line ID",
              "type": "scheme",
            },
            Object {
              "id": "mailto",
              "name": "Email Address",
              "type": "scheme",
            },
            Object {
              "id": "tel",
              "name": "Phone Number",
              "type": "scheme",
            },
            Object {
              "id": "telegram",
              "name": "Telegram ID",
              "type": "scheme",
            },
            Object {
              "id": "twitterid",
              "name": "Twitter ID",
              "type": "scheme",
            },
            Object {
              "id": "wechat",
              "name": "Wechat ID",
              "type": "scheme",
            },
            Object {
              "id": "whatsapp",
              "name": "Whatsapp Number",
              "type": "scheme",
            },
            Object {
              "id": "viber",
              "name": "Viber ID",
              "type": "scheme",
            },
          ]
        }
        assets={
          Object {
            "items": Object {},
            "type": "field",
          }
        }
        entry={
          Object {
            "value": Object {
              "id": "favorite_color",
              "name": null,
              "type": "field",
            },
          }
        }
        name="Contact Field"
        onChange={[Function]}
        searchable={true}
        sortFunction={[Function]}
        styles={
          Object {
            "control": [Function],
            "indicatorsContainer": [Function],
            "input": [Function],
            "menuList": [Function],
            "option": [Function],
            "singleValue": [Function],
            "valueContainer": [Function],
          }
        }
      />
    </div>
  </div>
  <CaseList
    cases={Array []}
    data-spec="cases"
    onCasesUpdated={[Function]}
  />
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`FieldRouterForm updates should save changes 1`] = `
Object {
  "cases": Array [
    Object {
      "categoryName": "Red",
      "kase": Object {
        "arguments": Array [
          "red",
        ],
        "type": "has_any_word",
      },
    },
    Object {
      "categoryName": "Red",
      "kase": Object {
        "arguments": Array [
          "maroon",
        ],
        "type": "has_any_word",
      },
    },
    Object {
      "categoryName": "Green",
      "kase": Object {
        "arguments": Array [
          "green",
        ],
        "type": "has_any_word",
      },
    },
  ],
  "field": Object {
    "value": Object {
      "id": "favorite_color",
      "name": null,
      "type": "field",
    },
  },
  "resultName": Object {
    "value": "Favorite Color",
  },
  "valid": true,
}
`;

exports[`FieldRouterForm updates should save changes 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        },
        Object {
          "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
        },
        Object {
          "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "@contact.fields.favorite_color",
              "red",
            ],
            "category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "@contact.fields.favorite_color",
              "maroon",
            ],
            "category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "type": "has_any_word",
          },
          Object {
            "arguments": Array [
              "@contact.fields.favorite_color",
              "green",
            ],
            "category_uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
            "type": "has_any_word",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "name": "Red",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
          Object {
            "exit_uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
            "name": "Green",
            "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
          },
          Object {
            "exit_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
            "name": "Other",
            "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
          },
        ],
        "default_category_uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
        "result_name": "Favorite Color",
        "type": "switch",
      },
      "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
    },
    "ui": Object {
      "config": Object {
        "router": Object {
          "cases": Object {},
          "operand": "@contact.fields.favorite_color",
          "operandAsset": Object {
            "id": "favorite_color",
            "name": null,
            "type": "field",
          },
        },
      },
      "position": null,
      "type": "split_by_contact_field",
    },
  },
]
`;
